modules/fdb/examples/sqlite-to-fdb.rs

```diff
@@ -1,45 +1,214 @@
-use assembly_fdb::{common::ValueType, core::Field, mem::Database, store};
+use assembly_fdb::{common::Latin1String, common::ValueType, core::Field, mem::Database, store};
 use color_eyre::eyre::{self, eyre, WrapErr};
 use mapr::Mmap;
 use rusqlite::{types::ValueRef, Connection};
 use std::{fmt::Write, fs::File, io::BufWriter, path::PathBuf, time::Instant};
 use structopt::StructOpt;
 
 #[derive(StructOpt)]
-/// Fills a template FDB file (see template-fdb.rs) with rows from a sqlite database
+#[structopt(name = "sqlite-to-fdb")]
+#[structopt(author = "zaop")]
+/// Convert an SQLite database to FDB. By default, type information from the SQLite DB is used; if unavailable, you can specify the target
+/// datatypes through a 'template' FDB file with `--template` (see template-fdb.rs).
 struct Options {
-    /// "Template" FDB file, the table structure from this will be used for the output
-    template: PathBuf,
-    /// Input sqlite database
+    /// Input SQLite file
     src: PathBuf,
     /// Output FDB file
     dest: PathBuf,
+    /// Optional: an FDB file containing tables with correct columns but no rows used to determine type information
+    #[structopt(long)]
+    template: Option<PathBuf>,
 }
 
 fn main() -> eyre::Result<()> {
     color_eyre::install()?;
     let opts = Options::from_args();
-    let start = Instant::now();
-
-    println!("Converting database, this may take a few seconds...");
 
-    // fdb template
-    let template_file = File::open(&opts.template)
-        .wrap_err_with(|| format!("Failed to open fdb template '{}'", opts.template.display()))?;
-    let mmap = unsafe { Mmap::map(&template_file)? };
-    let buffer: &[u8] = &mmap;
-    let template_db = Database::new(buffer);
+    let start = Instant::now();
 
     // sqlite input
+    if !opts.src.exists() {
+        return Err(eyre!("SQLite file '{}' not found", opts.src.display()));
+    }
     let conn = Connection::open_with_flags(&opts.src, rusqlite::OpenFlags::SQLITE_OPEN_READ_ONLY)
-        .wrap_err_with(|| format!("Failed to open sqlite file '{}'", opts.src.display()))?;
+        .wrap_err_with(|| format!("Failed to open SQLite file '{}'", opts.src.display()))?;
+
+    // this .unwrap() won't error because we just checked that the file exists and therefore it has a name
+    let src_stem = opts.src.file_stem().unwrap();
+    let mut dest_path = opts.dest.clone();
+
+    // check if destination path is a directory
+    if dest_path.is_dir() {
+        // save to source file name but with .fdb extension
+        dest_path = dest_path.join(src_stem).with_extension("fdb");
+        if dest_path.exists() {
+            return Err(eyre!(
+                "A directory was specified as output path, and in this directory the default filename '{}.fdb' is already in use.",
```

Comment:
Technically, this could be a question to overwrite (with a `-f` or `--force` flag override) instead of an error. Again, just a suggestion, not necessary to get this merged

Reply:
That way, you can also move this check out of the fallback for directories into and `if out_path.is_file()` or so

Reply:
I thought about doing something like this but the target filename inside the directory could also be another directory, in which case it'd either have to create a file inside _that_ folder (recursively) or delete the entire directory tree from there on down to make place for this file. Or still throw an error for directories, but use the `--force` flag only to override a file. But that feels a bit inconsistent. So I think it makes sense to simply require the user to specify the full, correct output filename to override an existing file (which currently happens automatically if a valid filename is supplied). What do you think?

Reply:
IMO if someone managed to create a folder with an fdb extension, they should just get an error. So yeah, if it's a file ask to overwrite, otherwise just bail.

Reply:
I think what feels wrong about this to me is that it's a fallback mechanism - if someone accidentally types a folder name, the program still tries to find a way to complete its task by creating a file inside that folder. Then adding a `--force` to intentionally overwrite an existing file inside that folder seems weird. If you were intentionally targeting a file in that folder, you would specify the output file you want, rather than only the folder name and leaving it to the program to come up with a filename.

Reply:
I guess I'm coming from the perspective that I never want to accidentally overwrite a file so it'd be weird not to have that check for both cases if we have it at all.

One reason why people might like to use the directory name is that it autocompletes in a Shell, whether you already have a file or not. And you get the filename match for free, which is nice.

Reply:
Okay yeah I agree that it'd be better to have that check be consistent. I could make it behave the same as `cp` and `mv` - overwrite by default if a file exists, with options that can be specified to make it either interactive or default to not overwriting.

Reply:
Yes, that sounds good!

Reply:
Done; `-f` overwrites, `-n` does not overwrite, and `-i` prompts the user before overwriting. (`-f` is the default but you can still specify the option to be more explicit if desired.)
Now of course it would be good to give the other tools the same behaviour but that's something for a separate PR.